pkg/controller/bastion/actuator_delete.go

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	azureclient "github.com/gardener/gardener-extension-provider-azure/pkg/azure/client"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	ctrlerror "github.com/gardener/gardener/extensions/pkg/controller/error"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (a *actuator) Delete(ctx context.Context, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	logger := a.logger.WithValues("bastion", client.ObjectKeyFromObject(bastion), "operation", "delete")
+
+	var factory = azureclient.NewAzureClientFactory(a.client)
+
+	opt, err := DetermineOptions(bastion, cluster)
+	if err != nil {
+		return err
+	}
+
+	err = removeBastionInstance(ctx, factory, logger, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove bastion instance: %w", err)
+	}
+
+	deleted, err := isInstanceDeleted(ctx, factory, opt)
```

Comment:
Is this just a double check to be sure that the machine is really gone?

Reply:
yes, sometimes instance takes a long time to remove or other networks/API issues.  make sure the instances had removed before deleting other dependent resources eg nic, public ip etc..  
https://github.com/tedteng/gardener-extension-provider-azure/blob/d5f6e908685bf64b297dd11ab9e6e5568d8751ee/pkg/controller/bastion/actuator_delete.go#L48-L58
 ```
	deleted, err := isInstanceDeleted(ctx, factory, opt)
	if err != nil {
		return fmt.Errorf("failed to check for bastion instance: %w", err)
	}

	if !deleted {
		return &ctrlerror.RequeueAfterError{
			RequeueAfter: 10 * time.Second,
			Cause:        errors.New("bastion instance is still deleting"),
		}
	}
```